src/backend/distributed/utils/meta_visibility.c

```diff
@@ -0,0 +1,700 @@
+/*
+ * citus_meta_visibility.c
+ *
+ * Implements the functions for hiding citus objects.
+ *
+ * Copyright (c) Citus Data, Inc.
+ */
+
+#include "postgres.h"
+#include "miscadmin.h"
+
+#include "access/genam.h"
+#include "catalog/pg_aggregate.h"
+#include "catalog/pg_am.h"
+#include "catalog/pg_attribute.h"
+#include "catalog/pg_attrdef.h"
+#include "catalog/pg_constraint.h"
+#include "catalog/pg_class.h"
+#include "catalog/pg_depend.h"
+#include "catalog/pg_enum.h"
+#include "catalog/pg_event_trigger.h"
+#include "catalog/pg_language.h"
+#include "catalog/pg_namespace.h"
+#include "catalog/pg_proc.h"
+#include "catalog/pg_rewrite.h"
+#include "catalog/pg_sequence.h"
+#include "catalog/pg_statistic.h"
+#include "catalog/pg_trigger.h"
+#include "catalog/pg_ts_config.h"
+#include "catalog/pg_ts_dict.h"
+#include "catalog/pg_ts_template.h"
+#include "catalog/pg_type.h"
+#include "catalog/namespace.h"
+#include "commands/defrem.h"
+#include "commands/extension.h"
+#include "common/hashfn.h"
+#include "distributed/citus_meta_visibility.h"
+#include "distributed/commands.h"
+#include "distributed/metadata_cache.h"
+#include "distributed/listutils.h"
+#include "distributed/log_utils.h"
+#include "nodes/makefuncs.h"
+#include "nodes/nodeFuncs.h"
+#include "nodes/parsenodes.h"
+#include "utils/builtins.h"
+#include "utils/fmgroids.h"
+#include "utils/lsyscache.h"
+#include "utils/syscache.h"
+
+/* memory context for allocating DependentObjects */
+static MemoryContext DependentObjectsContext = NULL;
+
+static Node * CreateCitusDependentObjectExpr(int pgMetaTableVarno, int pgMetaTableOid);
+static List * GetFuncArgs(int pgMetaTableVarno, int pgMetaTableOid);
+
+PG_FUNCTION_INFO_V1(is_citus_depended_object);
+
+/*
+ * is_citus_depended_object a wrapper around IsCitusDependentObject, so
+ * see the details there.
+ */
+Datum
+is_citus_depended_object(PG_FUNCTION_ARGS)
+{
+	CheckCitusVersion(ERROR);
+
+	Oid metaTableId = PG_GETARG_OID(0);
+	if (!OidIsValid(metaTableId))
+	{
+		/* we cannot continue without valid meta table oid */
+		PG_RETURN_BOOL(false);
+	}
+
+	bool dependsOnCitus = false;
+
+	Oid objectId = PG_GETARG_OID(1);
+
+	DependentObjectsContext =
+		AllocSetContextCreate(
+			CurrentMemoryContext,
+			"Dependent Objects Context",
+			ALLOCSET_DEFAULT_SIZES);
+
+	MemoryContext oldContext = MemoryContextSwitchTo(DependentObjectsContext);
+
+	switch (metaTableId)
+	{
+		/* handle meta objects that can be found in pg_depend */
+		case RelationRelationId:
```

Comment:
In general, this code seems to be specific per object type.

Is there a possibility to use `PgGetObjectAddress()` function such that we have a generic implementation?